File: app/src/main/java/it/ibashkimi/lockscheduler/BaseActivity.java
package it.ibashkimi.lockscheduler;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.Resources;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;

import it.ibashkimi.support.design.preference.Themes;
import it.ibashkimi.support.design.utils.ThemeUtils;


public class BaseActivity extends AppCompatActivity {

    private static final String PREFS_NAME = "prefs";
    private static final int DEFAULT_THEME = Themes.Theme.APP_THEME_DAYNIGHT_INDIGO;

    private SharedPreferences mSharedPrefs;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        mSharedPrefs = getSharedPreferences("prefs", Context.MODE_PRIVATE);
        ThemeUtils.applyDayNightMode(this, mSharedPrefs.getString("night_mode", "auto"));
        @Themes.Theme int themeId = mSharedPrefs.getInt("theme", DEFAULT_THEME);
        ThemeUtils.applyTheme(this, themeId);
        /*int theme = Themes.resolveTheme(themeId);
        int primaryDarkColor = ThemeUtils.getColorFromAttribute()*/
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            boolean coloredNavBar = mSharedPrefs.getBoolean("colored_navigation_bar", false);
            if (coloredNavBar) {
                getWindow().setNavigationBarColor(
                        ThemeUtils.getColorFromAttribute(this, R.attr.colorPrimaryDark));
            }
        }
        super.onCreate(savedInstanceState);
    }

    protected SharedPreferences getSharedPreferences() {
        if (mSharedPrefs == null) {
            mSharedPrefs = getSharedPreferences("prefs", Context.MODE_PRIVATE);
        }
        return mSharedPrefs;
    }
}

File: app/src/main/java/com/ibashkimi/lockscheduler/ui/BaseActivity.kt
package com.ibashkimi.lockscheduler.ui

import android.app.Activity
import android.content.Context
import android.content.SharedPreferences
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import com.ibashkimi.lockscheduler.data.prefs.AppPreferencesHelper


abstract class BaseActivity : AppCompatActivity() {

    private val preferencesName = AppPreferencesHelper.PREFERENCES_NAME

    val preferences: SharedPreferences by lazy {
        getSharedPreferences(preferencesName, Context.MODE_PRIVATE)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        if (savedInstanceState == null) {
            val nightMode = preferences.getString("night_mode", null) ?: "system_default"
            applyGlobalNightMode(nightMode)
        }

        super.onCreate(savedInstanceState)
    }
}

fun applyGlobalNightMode(nightMode: String) {
    AppCompatDelegate.setDefaultNightMode(
        when (nightMode) {
            "on" -> AppCompatDelegate.MODE_NIGHT_YES
            "system_default" -> AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM
            "battery_saver" -> AppCompatDelegate.MODE_NIGHT_AUTO_BATTERY
            "off" -> AppCompatDelegate.MODE_NIGHT_NO
            else -> throw IllegalArgumentException("Invalid night mode $nightMode.")
        }
    )
}

fun Activity.applyNightMode(nightMode: String) {
    applyGlobalNightMode(nightMode)
    recreate()
}
